rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/TuplespaceAlg.scala

```diff
@@ -0,0 +1,81 @@
+package coop.rchain.rholang.interpreter.storage
+
+import cats.Parallel
+import cats.effect.Sync
+import coop.rchain.models.Channel.ChannelInstance.Quote
+import coop.rchain.models._
+import coop.rchain.models.TaggedContinuation.TaggedCont.ParBody
+import coop.rchain.rholang.interpreter.Dispatch
+import coop.rchain.rholang.interpreter.errors.ReduceError
+import coop.rchain.rspace.pure.PureRSpace
+import cats.implicits._
+import coop.rchain.models.rholang.implicits._
+import coop.rchain.rholang.interpreter.storage.implicits._
+
+trait TuplespaceAlg[F[_]] {
+  def produce(chan: Channel, data: ListChannelWithRandom, persistent: Boolean): F[Unit]
+  def consume(binds: Seq[(BindPattern, Quote)], body: ParWithRandom, persistent: Boolean): F[Unit]
+}
+
+object TuplespaceAlg {
+  def rspaceTuplespace[F[_], M[_]](
+      pureRSpace: PureRSpace[F,
+                             Channel,
+                             BindPattern,
+                             ListChannelWithRandom,
+                             ListChannelWithRandom,
+                             TaggedContinuation],
+      dispatcher: => Dispatch[F, ListChannelWithRandom, TaggedContinuation])(
+      implicit F: Sync[F],
+      P: Parallel[F, M]): TuplespaceAlg[F] = new TuplespaceAlg[F] {
+    override def produce(channel: Channel,
+                         data: ListChannelWithRandom,
+                         persistent: Boolean): F[Unit] = {
+      // TODO: Handle the environment in the store
+      def go(res: Option[(TaggedContinuation, Seq[ListChannelWithRandom])]): F[Unit] =
+        res match {
```

Comment:
instead of "manual" matching the Option you could use getOrElse. I find it more readable than doing the match but it's a matter of preference.